elasticsearch/src/main/scala/akka/stream/alpakka/elasticsearch/ElasticsearchSinkStage.scala

```diff
@@ -0,0 +1,175 @@
+                                                in: Inlet[IncomingMessage[T]],
+                                                promise: Promise[Done])
+    extends GraphStageLogic(shape) {
+
+  private val state = new AtomicReference[State](Idle)
+  private val buffer = new util.concurrent.ConcurrentLinkedQueue[IncomingMessage[T]]()
+  private val failureHandler = getAsyncCallback[Throwable](handleFailure)
+  private val sentHandler = getAsyncCallback[Unit](_ => handleSent())
+
+  protected def convert(value: T): JsObject
+
+  override def preStart(): Unit =
+    pull(in)
+
+  private def tryPull(): Unit =
+    if (buffer.size < settings.bufferSize && !isClosed(in) && !hasBeenPulled(in)) {
+      pull(in)
+    }
+
+  private def handleFailure(exception: Throwable): Unit = {
+    failStage(exception)
+    promise.tryFailure(exception)
+  }
+
+  private def handleSuccess(): Unit = {
+    completeStage()
+    promise.trySuccess(Done)
+  }
+
+  private def handleSent(): Unit =
+    tryPull()
+
+  setHandler(in,
+    new InHandler {
+    override def onPush(): Unit = {
+      val message = grab(in)
+      buffer.add(message)
+
+      state.get match {
+        case Idle => {
+          state.set(Sending)
+          val messages = (1 to settings.bufferSize).flatMap { _ =>
+            Option(buffer.poll())
+          }
+          sendBulkUpdateRequest(messages)
+        }
+        case _ => ()
+      }
+
+      tryPull()
+    }
+
+    override def onUpstreamFailure(exception: Throwable): Unit =
+      handleFailure(exception)
+
+    override def onUpstreamFinish(): Unit =
+      state.get match {
+        case Idle => handleSuccess()
+        case Sending => state.set(Finished)
+        case Finished => ()
+      }
+
+    private def sendBulkUpdateRequest(messages: Seq[IncomingMessage[T]]): Unit =
+      try {
+        val json = messages.map { message =>
+          s"""{"index": {"_index": "${indexName}", "_type": "${typeName}"${message.id.map { id =>
+               s""", "_id": "${id}""""
+             }.getOrElse("")}}
+          |${convert(message.source).toString}""".stripMargin
+        }.mkString("", "\n", "\n")
+
+        client.performRequestAsync(
+          "POST",
+          "_bulk",
+          Map[String, String]().asJava,
+          new StringEntity(json),
+          new ResponseListener {
+            override def onFailure(exception: Exception): Unit = failureHandler.invoke(exception)
+
+            override def onSuccess(response: Response): Unit = {
+              val messages = (1 to settings.bufferSize).flatMap { _ =>
+                Option(buffer.poll())
+              }
+
+              if (messages.isEmpty) {
+                state.get match {
+                  case Finished => handleSuccess()
+                  case _ => state.set(Idle)
+                }
+              } else {
+                sendBulkUpdateRequest(messages)
+              }
+
+              sentHandler.invoke(())
+            }
+          }
```

Comment:
Hmm that's still quite dangerous to do because you can run in all kinds of race conditions. The one that I can see is the `state.set(Idle)` case. When the stage fills the buffer after the `val messages` in line 150 but before line 155 is executed, then the state will be set to `Idle` but `tryPull` won't do anything and the stream is locked because it will never be woken up again. Not very likely but still possible.

Once you do all handling in the async callback you can remove all other synchronization like the `AtomicReferences` and the `ConcurrentLinkedHashmap`.